File: backend/app/schemas/gps.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


GPS_PRESENCE_STATUSES = {"not_checkable", "missing", "partial", "matched", "mismatch"}


class GpsLocationPointCreate(BaseModel):
    person_id: int | None = None
    captured_at: datetime
    latitude: float = Field(ge=-90, le=90)
    longitude: float = Field(ge=-180, le=180)
    accuracy_meters: float | None = Field(default=None, ge=0)
    source: str | None = Field(default="mobile", max_length=40)
    device_id: str | None = Field(default=None, max_length=120)


class GpsLocationPointRead(BaseModel):
    id: int
    person_id: int
    captured_at: datetime
    latitude: float
    longitude: float
    accuracy_meters: float | None = None


class GpsRecentLocationPointRead(BaseModel):
    id: int
    person_id: int
    person_name: str
    captured_at: datetime
    planned_site_id: int | None = None
    planned_site_label: str | None = None
    plausibility_status: str
    distance_to_planned_site_m: float | None = None
    geofence_radius_m: int | None = None


class GpsPresenceStatusRead(BaseModel):
    status: str
    matched_points: int
    total_points: int
    reason: str

File: frontend/src/types/gps.ts
export type GpsLocationPointCreate = {
  captured_at: string;
  latitude: number;
  longitude: number;
  accuracy_meters?: number | null;
  source?: "mobile";
  device_id?: string | null;
};

export type GpsLocationPointRead = {
  id: number;
  person_id: number;
  captured_at: string;
  latitude: number;
  longitude: number;
  accuracy_meters: number | null;
};

export type GpsRecentLocationPoint = {
  id: number;
  person_id: number;
  person_name: string;
  captured_at: string;
  planned_site_id: number | null;
  planned_site_label: string | null;
  plausibility_status: "matched" | "mismatch" | "not_checkable" | "partial" | "missing";
  distance_to_planned_site_m: number | null;
  geofence_radius_m: number | null;
};
